### PR TITLE
[SV][ExportVerilog] Add sv.struct_field_inout op

### PR DESCRIPTION
**Pull request by uenoku**
This commit adds an new operation `sv.struct_field_inout` to create
inout memory for hw::StructType. Potentially, this will be used to
lower firrtl bundle types.

### CODE REVIEW

**Review by lattner — APPROVED** [review]
Nice! 

Did you consider making the MLIR syntax more struct like?  Instead of `sv.struct_field_inout %a["b"] ` you could use `sv.struct_field_inout %a."b" ` or even `sv.struct_field_inout %a.b` using the `parseKeywordOrString` logic.

**Comment by uenoku** [comment]
> Did you consider making the MLIR syntax more struct like? Instead of sv.struct_field_inout %a["b"] you could use sv.struct_field_inout %a."b" or even sv.struct_field_inout %a.b using the parseKeywordOrString logic.

Thanks! I'll try it.

**Comment by uenoku** [comment]
> Did you consider making the MLIR syntax more struct like? Instead of sv.struct_field_inout %a["b"] you could use sv.struct_field_inout %a."b" or even sv.struct_field_inout %a.b using the parseKeywordOrString logic.

It seems using a dot as keywards is prohibited at the lexer level now. So I feel it is necessary to modify MLIR lexer.

**Comment by darthscsi** [comment]
> > Did you consider making the MLIR syntax more struct like? Instead of sv.struct_field_inout %a["b"] you could use sv.struct_field_inout %a."b" or even sv.struct_field_inout %a.b using the parseKeywordOrString logic.
> 
> It seems using a dot as keywards is prohibited at the lexer level now. So it is necessary to modify MLIR lexer. I'm not sure this is an easy change. Since we are currently allowed to use a dot in SSA name, we can't distinguish `%a.b` whether it is a field access or just name.

I believe that's why the other struct ops don't use dots.

File: include/circt/Dialect/SV/SVInOutOps.td [hunk 1]
```diff
@@ -176,3 +176,30 @@ def IndexedPartSelectInOutOp : SVOp<"indexed_part_select_inout",
...
+def InOutStructType
+  : Type<CPred<"getInOutElementType($_self).isa<hw::StructType>()">,
+         "an inout type with struct field", "::circt::hw::InOutType">;
+
+def StructFieldInOutOp : SVOp<"struct_field_inout",
+                              [NoSideEffect, InferTypeOpInterface]> {
+  let summary = "Create an subfield inout memory to produce an inout element.";
+  let description = "See SV Spec 7.2.";
+
+  let arguments = (ins InOutStructType:$input, StrAttr:$field);
```
Comment (darthscsi) [inline include/circt/Dialect/SV/SVInOutOps.td:189]: We had been converting to using integer indexes into structs to reduce verification.

Comment (darthscsi) [reply include/circt/Dialect/SV/SVInOutOps.td:189]: but I guess that's only in FIRRTL still, so nevermind.

File: include/circt/Dialect/SV/SVInOutOps.td [hunk 2]
```diff
@@ -176,3 +176,30 @@ def IndexedPartSelectInOutOp : SVOp<"indexed_part_select_inout",
...
+  let results = (outs InOutType:$result);
+
+  let assemblyFormat =
+    "$input `[` $field `]` attr-dict `:` type($input)";
```
Comment (darthscsi) [inline include/circt/Dialect/SV/SVInOutOps.td:193]: This is consistent with the other struct ops, so don't make it different for this one.  Changing the syntax for struct indexing should be a different PR against all the ops at once, if we want to do that,  Which is to say, this is fine for this PR.

**Review by darthscsi — APPROVED** [review]
LGTM

File: include/circt/Dialect/SV/SVInOutOps.td (hunk 2 above)
Comment (uenoku) [reply include/circt/Dialect/SV/SVInOutOps.td:193]: > Changing the syntax for struct indexing should be a different PR against all the ops at once, if we want to do that, Which is to say, this is fine for this PR.

Yes, agreed.
